zebrad/tests/acceptance.rs

```diff
@@ -1052,3 +1055,157 @@ async fn tracing_endpoint() -> Result<()> {
+
+    // From another folder create a configuration with the same endpoint.
+    // `metrics.endpoint_addr` will be the same in the 2 nodes.
+    // But they will have different Zcash listeners (auto port) and states (ephemeral)
+    let dir2 = TempDir::new("zebrad_tests")?.with_config(&mut config)?;
+
+    check_config_conflict(dir1, regex1.as_str(), dir2, PORT_IN_USE_ERROR)?;
+
+    Ok(())
+}
+
+/// Start 2 zebrad nodes using the same tracing listener port, but different
+/// state directories and Zcash listener ports. The first node should get
+/// exclusive use of the port. The second node will panic with the Zcash tracing
+/// conflict hint added in #1535.
+#[test]
+fn zcash_tracing_conflict() -> Result<()> {
+    zebra_test::init();
+
+    // [Note on port conflict](#Note on port conflict)
+    let port = random_known_port();
+    let listen_addr = format!("127.0.0.1:{}", port);
+
+    // Write a configuration that has our created tracing endpoint_addr
+    let mut config = default_test_config()?;
+    config.tracing.endpoint_addr = Some(listen_addr.parse().unwrap());
+    let dir1 = TempDir::new("zebrad_tests")?.with_config(&mut config)?;
+    let regex1 = format!(r"Opened tracing endpoint at {}", listen_addr);
+
+    // From another folder create a configuration with the same endpoint.
+    // `tracing.endpoint_addr` will be the same in the 2 nodes.
+    // But they will have different Zcash listeners (auto port) and states (ephemeral)
+    let dir2 = TempDir::new("zebrad_tests")?.with_config(&mut config)?;
+
+    check_config_conflict(dir1, regex1.as_str(), dir2, PORT_IN_USE_ERROR)?;
+
+    Ok(())
+}
+
+/// Start 2 zebrad nodes using the same state directory, but different Zcash
+/// listener ports. The first node should get exclusive access to the database.
+/// The second node will panic with the Zcash state conflict hint added in #1535.
+#[test]
+fn zcash_state_conflict() -> Result<()> {
+    zebra_test::init();
+
+    // A persistent config has a fixed temp state directory, but asks the OS to
+    // automatically choose an unused port
+    let mut config = persistent_test_config()?;
+    let dir_conflict = TempDir::new("zebrad_tests")?.with_config(&mut config)?;
+
+    check_config_conflict(
+        dir_conflict.path(),
+        "Opened Zebra state cache",
+        dir_conflict.path(),
+        LOCK_FILE_ERROR,
+    )?;
+
+    Ok(())
+}
+
+/// Launch a node in `first_dir`, wait a few seconds, then launch a node in
+/// `second_dir`. Check that the first node's stdout contains
+/// `first_stdout_regex`, and the second node's stderr contains
+/// `second_stderr_regex`.
+fn check_config_conflict<T, U>(
+    first_dir: T,
+    first_stdout_regex: &str,
+    second_dir: U,
+    second_stderr_regex: &str,
+) -> Result<()>
+where
+    T: ZebradTestDirExt,
+    U: ZebradTestDirExt,
+{
+    // By DNS issues we want to skip all port conflict tests on macOS by now.
+    // Follow up at #1631
+    if !cfg!(target_os = "macos") {
```

Comment:
one style suggestion, in cases like this I think its nicer to structure the filter as an early exit, that way you dont end up indenting the entire function.

```rust
if cfg!(target_os = "macos") {
    return Ok(());
}

// the content of the old if statement ...
```

Reply:
Definitely better, applied at https://github.com/ZcashFoundation/zebra/pull/1535/commits/7ee9149c449f0b29127ef097e7973d89b2f5a267 Thanks!